### ISSUE TITLE
twisted.conch.ssh.userauth is poorly tested.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2682]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| z3p reported|
|-|-|
|Trac ID|trac#2682|
|Type|defect|
|Created|2007-05-29 05:32:30Z|
|Branch|https://github.com/twisted/twisted/tree/userauth-2682-7|

The reasons are too large to enumerate, but twisted.conch.ssh.userauth needs better tests.

Attachments:

* [userauth-imports-2682.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/4a3/4a3669d650b1f3aa26bd356022e1546980151de3/405e505ee9a53b1fadefb456dacd7dd97764610d.patch) (7742 bytes) - added by thijs on 2008-07-13 09:55:30Z - Adding patch that removes the remaining relative imports

<details><summary>Searchable metadata</summary>

```
trac-id__2682 2682
type__defect defect
reporter__z3p z3p
priority__high high
milestone__ 
branch__branches_userauth_2682_7 branches/userauth-2682-7
branch_author__ 
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__soc2007 soc2007
time__1180416750000000 1180416750000000
changetime__1244293782000000 1244293782000000
version__None None
owner__ 
cc__therve cc__thijs cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|z3p set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|z3p removed owner<br>z3p set status to `new`|
|-|-|

Ready for review in branches/userauth-2862.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve commented|
|-|-|

 * The ```import ()``` syntax is not supported under python2.3
 * LoopbackTestCase.test_loopback has a problem when run 3 times, but I wasn't able to spot it. To reproduce, run ```trial -u twisted.conch.test.test_userauth.LoopbackTestCase.test_loopback```. Overall, this test looks very complicated.

That's it for now.

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|z3p commented|
|-|-|

I'm not sure how to make it less complicated than it is, but I've fixed the problem where it would crash after 3 runs.  I've also fixed the import () syntax.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to z3p<br>@exarkun set status to `new`|
|-|-|

Hmmm.  Changes to checkers.py probably look good (although that class could use some docstrings), but I'm not really sure why SSHProtocolChecker exists in the first place.  At most, it seems like a unit test utility class.  Does it belong somewhere beneath the twisted.conch.test package instead?

In userauth.py, the relative imports should probably be replaced with absolute imports (at least the ones which are changing, but I don't think there's a reason not to fix them all).  Also please add some whitespace around + in ```SSHUserAuthClient._setNewPass```.  In ```SSHUserAuthServer.ssh_USERAUTH_REQUEST```, there is a change to fix what I think was a NameError for ConchError - calling self._ebBadAuth with error.ConchError(...) instead of ConchError(...) - but there's no test coverage for this line.  SSHUserAuthServer and SSHUserAuthClient are missing class docstrings (include attribute docs).

In test_userauth.py, the stub transport and userauth classes are defined with simple suites - put the ```pass``` statements on the next line (better, replace them with docstrings).  Also, ClientUserAuth is missing a bunch of docstrings, and getPrivateKey has a line >80 cols.  I'd expand "KI" in "test_removeKIIfUnencrypted".  test_loginTimeout and test_canceLoginTimeout should both use a fake clock to make their operation deterministic - see twisted.internet.task.Clock.  ```_``` function in ```test_USERAUTH_SUCCESS``` should be renamed.  ```test_keyboard_interactive``` should be renamed ```test_keyboardInteractive```.  Some whitespace before LoopbackTestCase would be nice. :)  The nested Factory and Service classes are missing docstrings.  ```_``` function in makeConnection in setUp in test_loopback should be renamed.

test_loopback is still doing a lot of stuff.  Much of this could be dropped if the test didn't try to implement so much of the SSH transport interface (which clearly isn't the same as ITCPTransport).  Is there a reason an existing conch class can't be wrapped around the underlying loopback transport and that object exposed to the instances in the test method?  If this could be done, it should remove the need for the entire setUp function, at which point the test method is almost reasonably sized.

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|z3p removed owner|
|-|-|

Re: SSHProtocolChecker: it's been in checkers because it's (potentially) useful as a checker which needs multiple authentications to succeed.

The other changes have been made and I'm putting this back up for review.

**Comment by twisted-trac** [comment]
|[<img alt="Peaker's avatar" src="https://avatars.githubusercontent.com/u/48899?s=50" width="50" height="50">](https://github.com/Peaker)<a name="note_8"></a>|@Peaker commented|
|-|-|

Replying to [z3p](#note_2):
> Ready for review in branches/userauth-2862.

A. Is that a typo here or in the branch name (2862 instead of 2682)?
B. Is this branch still the right place to review?

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|z3p commented|
|-|-|

Replying to [Peaker](#note_8):
> Replying to [z3p](#note_2):
> > Ready for review in branches/userauth-2862.
>
> A. Is that a typo here or in the branch name (2862 instead of 2682)?
> B. Is this branch still the right place to review?

Yes, that's a typo.  The correct branch is userauth-2682.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_10"></a>|@radix set owner to z3p|
|-|-|

1. Indentation in userauth.py is slightly incorrect; second two lines should be indented one more space.

+            b = (NS(self.transport.sessionID) + chr(MSG_USERAUTH_REQUEST) +
+                NS(self.user) + NS(self.instance.name) + NS('publickey') +
+                '\xff' + NS(keyType) + NS(publicKey))


2. Docstrings are missing from functions that have been changed. Notably, can you please add a docstring to SSHProtocolChecker (and/or requestAvatarId) which describes the entire process of authentication, including how areDone is used (because areDone looks dubious at best and really needs a good explanation if it's to stay).

3.     @type nextService: C{stR} in userauth.py

4.
    @ivar supportedAuthentications: A list of the supported authentication
        methods.
    @type supportedAuthentications: C{list}

List of what? I guess strings.

5. I can't read all of this at once, I need some ice cream. Thanks.

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|z3p set owner to @radix|
|-|-|

1-4 are taken care of.  As for 5, if you come visit me here in Northampton I'll buy you a cone. :)  Back up for review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to z3p|
|-|-|

I can sort of see what role SSHProtocolChecker is filling now.  It doesn't seem like a very nice one.  It also violates the interface required of checkers (by raising NotEnoughAuthentication) and doesn't work with concurrent logins for the same avatar identifier.  It's not new in the branch, so this doesn't need to hold up this ticket, I think, but it seems like it should be removed at some point, or at least drastically changed (I'm not sure what shape would make sense, though).  The only thing to note here is that the dependence of the new test_loopback on SSHProtocolChecker will make it harder to remove SSHProtocolChecker in the future.  AFAIK there is no other code in Twisted which uses it.

There is some trailing whitespace in files changed in this branch and some pyflakes warnings (not all of which are real problems, but some are) as well.  There's also a backslash line continuation which should be removed in test_loopback.

The rest looks okay.

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_13"></a>|z3p removed owner|
|-|-|

Ready for review in userauth-2682-2.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set owner to z3p|
|-|-|

userauth-2682-2 was created from a revision of trunk which had the [[#2679](https://github.com/twisted/twisted/issues/2679)](https://github.com/twisted/twisted/issues/2679) branch merged, so this code doesn't work merged to trunk at the moment.  There should probably be a -3 made after #2679 is re-merged.

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_15"></a>|z3p removed owner|
|-|-|

userauth-2682-2 merges to trunk just fine.  I added an extra test since the last review, for old getPrivateKey() behavior which the cftp tests required.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun set owner to z3p|
|-|-|

Lots of test failures on the debian-py2.4-select slave.

Mostly shaped like this:

```
[ERROR]: twisted.conch.test.test_keys.KeyTestCase.test_fromFile

Traceback (most recent call last):
  File "/srv/bb-slave/Run/slave/full2.4/Twisted/twisted/conch/test/test_keys.py", line 567, in test_fromFile
    self.assertEquals(keys.Key.fromFile(self.keyFile),
  File "/srv/bb-slave/Run/slave/full2.4/Twisted/twisted/conch/ssh/keys.py", line 58, in fromFile
    return Class.fromString(file(filename).read(), type, passphrase)
  File "/srv/bb-slave/Run/slave/full2.4/Twisted/twisted/conch/ssh/keys.py", line 84, in fromString
    return method(data)
  File "/srv/bb-slave/Run/slave/full2.4/Twisted/twisted/conch/ssh/keys.py", line 234, in _fromString_PRIVATE_LSH
    kd[name] = common.getMP(common.NS(data))[0]
  File "/srv/bb-slave/Run/slave/full2.4/Twisted/twisted/conch/ssh/common.py", line 91, in _fastgetMP
    mp.append(long(gmpy.mpz(data[c+4:c+l+4][::-1]+'\x00')))
exceptions.ValueError: string without NULL characters expected
```

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_17"></a>|z3p commented|
|-|-|

(In [22470]) Branching to 'userauth-2682-3'

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_18"></a>|z3p removed owner|
|-|-|

A quick merge-forward picked up a fix for that bug.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_19"></a>|@therve set owner to z3p|
|-|-|

 * in tryAuth, second and third `ConchError` are untested
 * in _cbFinishedAuth, the `ConchError` is untested
 * _pamConv `ConchError`s are untested
 * ssh_USERAUTH_INFO_RESPONSE use a bare expect, not tested either
 * _ebAuth is untested
 * ssh_USERAUTH_FAILURE both `ValueError`s are untested
 * _cbUserauthFailure `StopIteration` is untested
 * in ssh_USERAUTH_PK_OK, the case where the return value of self.signData is false is untested
 * in _cbGetPublicKey, the case where publicKey is not a Key is untested
 * in auth_password, the return 0 case is untested
 * in signData, the case where self.getPrivateKey returns None is untested.
 * getPublicKey is not tested. The comment in it should be removed.
 * getPrivateKey, getPassword and getGenericAnswers are untested.
 * the assert in ssh_USERAUTH_INFO_RESPONSE should be removed

Thanks!

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_20"></a>|z3p commented|
|-|-|

(In [22566]) checkpointing for the night

Refs [#2682](https://github.com/twisted/twisted/issues/2682)

review items this fixes:
# in tryAuth, second and third ConchError are untested
# in _cbFinishedAuth, the ConchError is untested
# _pamConv ConchErrors are untested
# ssh_USERAUTH_INFO_RESPONSE use a bare expect, not tested either
# _ebAuth is untested
# the assert in ssh_USERAUTH_INFO_RESPONSE should be removed

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_22"></a>|@therve commented|
|-|-|

There is a small conflict with trunk, it would be great if you could resolve it.

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_23"></a>|z3p commented|
|-|-|

(In [23057]) Branching to 'userauth-2682-4'

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_24"></a>|z3p commented|
|-|-|

Resolved in the new branch.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_25"></a>|@therve set owner to @therve|
|-|-|

 * key authentication with the conch client is broken
 * the bare except in ssh_USERAUTH_INFO_RESPONSE is unfortunate: it masks the wrong `ConchError` (it should be `error.ConchError`). The solution is maybe to use log.err in _ebBadAuth if the error is not `ConchError` (instead of `log.msg(reason.getTraceback())`)
 * trailing whitespaces in test_userauth
 * `ClientUserAuth.getPassword` has a useless blank line
 * SSHUserAuthServerTestCase docstring is not well formatted
 * many blank lines are missing. Sometimes there are a little bit too much :)
 * ssh_USERAUTH_SUCCESS has a commented line
 * the _pamDeferred attribute is not well managed in SSHUserAuthServer. It should take a default value, not be deleted when unused (rule: if you use hasattr, there is probably something wrong).
 * ssh_USERAUTH_PK_OK could be split in 3 smaller functions
 * auth_password is documented to return a bool, so it should use True and False. This is applicable to some other functions as well.
 * the code at the end of userauth should use locals() instead of importing userauth. It should delete the temporary variables too.

That's it for now!

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to z3p|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_29"></a>|@therve commented|
|-|-|

Thanks for the patch, but I don't think that'll help z3p much on these branches. Also, the status review should be given on the current branch addressing the problem, not on a patch on the current branch. If you want do help, getting in touch with z3p and applying the change directly could be great. Sorry about this.

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_30"></a>|z3p commented|
|-|-|

(In [24598]) Branching to 'userauth-2682-5'

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_31"></a>|z3p set status to `assigned`|
|-|-|

A change outside Conch (I imagine to Trial) has made one of my test-cases start failing.  LoopbackTestCase.test_loopback is failing because the warning isn't being raised in client.startService() anymore.  I'm not sure what to do about it; it's the only thing keeping me from putting this back up for review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_32"></a>|@exarkun commented|
|-|-|

It looks like the reason the warning isn't being emitted there might be that `SSHUserAuthClient.serviceStarted` isn't directly related to the deprecation; instead, it does some I/O, waits for a response from the server, then possibly does something that's deprecated.  I think it's probably loopback which changed to break this test, by not delivering data between connected protocols as quickly as it used to.  I suspect it'd be better to test for the warning in a much simpler more specific test, perhaps one which calls `tryAuth` or `auth_publickey` directly.

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_33"></a>|z3p commented|
|-|-|

I agree that testing that warning in a more specific test is a good idea.  How else can I get that warning (which I know will happen at some point) to not appear in the Trial output?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_34"></a>|@exarkun commented|
|-|-|

Use `twisted.trial.util.suppress` to create a suppression.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_35"></a>|@exarkun commented|
|-|-|

I was reflecting a bit further on my last comment.

Another solution which suggests itself is this.  If there is a more specific test (a _unit_ test, if you will ;) for the warning being emitted in this case, then there's really no reason for these other larger tests, tests primarily for other functionality, to exercise the deprecated code paths.  Perhaps instead of suppressing the warning in the tests, the implementation should be fixed so as not to emit it in these cases?

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_36"></a>|z3p commented|
|-|-|

(In [25049]) Branching to 'userauth-2682-6'

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_37"></a>|z3p removed owner<br>z3p set status to `new`|
|-|-|

Ready for review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_38"></a>|@exarkun set owner to z3p|
|-|-|

  1. Coding standard stuff
    1. update copyright date in modified files
    1. `SSHUserAuthServer.tryAuth` formats a string to pass to `error.ConchError` by passing a string as the right hand argument to % - it should pass a one tuple instead.
    1. `SSHProtocolChecker.requestAvatarId` needs a docstring
    1. So does `SSHProtocolChecker._cbGoodAuthentication`
    1. The formatting of the first line of `SSHProtocolChecker.areDone`'s docstring doesn't conform to the coding standard.
    1. A lot of `C{}` markup could be `L{}` markup instead.  For example, in the docstring of `SSHUserAuthServer`, `L{twisted.cred.portal.Portal}` would be good.
    1. `SSHUserAuthServer.serviceStopped` needs a docstring
  1. Incompatible changes
    1. `SSHProtocolChecker.areDone`'s signature changed.  The docstring of `areDone` encourages users to override the method, so old-style implementations should be supported (deprecated, if you want).  Using a new method name might be simpler than trying to inspect parameter count.  Up to you, though.
  1. Some untested codepaths in userauth.py
    1. in `SSHUserAuthServer.serviceStarted`, the checks for `'password'` and `'keyboard-interactive'` being in `self.supportedAuthentications` are only tested for the positive case, not the negative case.  Before those, the `i in self.interfaceToMethod` check is only tested for the positive case as well.
    1. in `SSHUserAuthClient.tryAuth`, the false case for the `if f:` conditional isn't tested.
    1. in `SSHUserAuthClient.ssh_USERAUTH_PK_OK` the false case for the `if func is not None:` conditional isn't tested.
  1. `SSHUserAuthClient.ssh_USERAUTH_FAILURE` might be much simplified by the use of `twisted.python.util.dsu`.
  1. `SSHProtocolChecker.areDone`'s docstring says "checkers must return the same type of avatar".  Does it mean "the same avatarId"?
  1. `SSHUserAuthServer.auth_password` misspells _authentication_ in its docstring
  1. all of `test_userauth` is pretty quick now, except for `LoopbackTestCase.test_loopback`.  I think that's because it's hitting the bad password delay.  It'd be great if this test didn't have to sleep for 1 second.
  1. `SSHUserAuthServer._cbFinishedAuth` sets `isAuthorized` on its transport.  This would be kind of a crummy API, except it doesn't actually seem to be an API for anything?  I don't see `isAuthorized` read or written anywhere else in conch.  I guess this is leftover cruft?
  1. `SSHUserAuthServer.authenticatedWith` is documented as a list... but of what?

When using an OpenSSH client against a Conch server, password authentication is broken.  However, it seems broken in trunk too.  If this is right and I'm not testing it wrong somehow, please file a new ticket.

The branch looks like it is getting pretty close.  Nice work.

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_39"></a>|z3p commented|
|-|-|

What would a test for the various negative cases look like?  Trying them and making sure an exception isn't raised?

Also, I just tried to connect to doc/conch/examples/sshsimpleserver.py with OpenSSH 5.1p1 and password authentication worked fine (both in the branch and in trunk).  What's not working for you?

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_40"></a>|z3p removed owner|
|-|-|

I added some whitebox tests for the negative cases.  Since I can't reproduce the password authentication issue and haven't gotten more information, I'm putting this back up for review.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_41"></a>|@therve commented|
|-|-|

Here we go:

```
try:
    import Crypto.Cipher.DES3, Crypto.Cipher.XOR
    from twisted.conch.ssh.common import NS
    from twisted.conch.checkers import SSHProtocolChecker
    from twisted.conch.ssh import keys, userauth, transport
    from twisted.conch.test import keydata
except ImportError:
    keys = None
```

This could lead to unnoticed skipped tests if there is an import error in any of the file unrelated to Crypto. Can you move the imports into an else clause?

`test_requestRaisesConchError` looks a bit fragile. I think it would be better to fire a test global Deferred with the failure in `mockEbBadAuth`, and use assertFailure on that Deferred.

In the same spirit, `test__pamConvErrors` and `test_tryAuthEdgeCases` should use assertFailure.

More later, I hope.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_42"></a>|@exarkun set owner to z3p|
|-|-|

  1. `SSHProtocolChecker.requestAvatarId`'s docstring starts with a typo
  1. `SSHProtocolChecker._ebGoodAuthentication`'s docstring has some ungrammar near its start - _if the a checker_
  1. `twisted.conch.test.test_userauth.ClientUserAuth.getPublicKey` documents and sets `usedKeys`, but this doesn't seem to be used by anything.
  1. `twisted.conch.test.test_userauth.LoopbackTestCase.test_loopback` has a suppression for _all_ deprecation warnings.  This seems a bit excessive.
  1. `SSHUserAuthServer._cbFinishedAuth` still sets `isAuthorized`.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_43"></a>|@exarkun commented|
|-|-|

Ah, one other thing, there are a lot of test failures in the branch.  These seem to be things which have since been fixed in trunk.  It would be good to merge the branch forward once more so as to be able to verify this, though.

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_44"></a>|z3p commented|
|-|-|

(In [26137]) Branching to 'userauth-2682-7'

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_45"></a>|z3p removed owner|
|-|-|

I addressed the comments from exarkun and therve, and there don't seem to be any test failures.  I think it's ready for review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_47"></a>|@exarkun set owner to z3p<br>@exarkun set status to `new`|
|-|-|

Some things I noticed during this review which it'd be great to get fixed separately from this set of changes:

  1. `twisted.conch.checkers.SSHPublicKeyDatabase._cbRequestAvatarId` doesn't have good coverage for the `if not validKey:` case nor for the exception handler or the final return statement.  Also, logging the error and returning a failure is probably redundant.  If the goal is to report the unexpected exception, it should probably be logged and then a /different/ failure returned (to normalize the type for whatever code invoked this).
  1. No test coverage for the final return statement in `twisted.conch.checkers.SSHProtocolChecker.requestAvatarId`.
  1. No test coverage for `twisted.conch.checkers.SSHProtocolChecker.areDone`.
  1. The logging going on in `twisted.conch.ssh.userauth.SSHUserAuthServer._ebBadAuth` is weird in a few ways:
    1. `reason.check(IgnoreAuthentication)` is better than `reason.type == IgnoreAuthentication`
    1. Failures should generally be `log.err()`'d, not `log.msg(str(...))`'d.
    1. If the wrong password was given, this is probably worth logging specifically without involving the string representation of a `Failure` instance or a traceback.  Something that is more easily recognizable as a user error, not a programming error.
  1. `pamauth` is imported in `twisted.conch.checkers` but not used (and there's no such module, I think).  I think this means pam authentication is broken in the server.  This doesn't seem to be new in this branch, though.

Other things to fix:

  1. _5. SSHUserAuthServer._cbFinishedAuth still sets isAuthorized._  I know this has been through a lot of reviews and you've done a huge amount of work on it, but it's really frustrating to see points from previous reviews go unaddressed (_twice_ in this case) when re-reviewing code.
  1. `twisted.conch.ssh.userauth.SSHUserAuthClient._cbGetPublicKey` and `_cbSignData` emit deprecation warnings without a version number.  They should say when the deprecation was first introduced - probably 9.0.
  1. `twisted.conch.test.test_userauth.SSHUserAuthServerTestCase.test_requestRaisesConchError` defines a `_callback` which is unused.  Also, it doesn't return the result of `assertFailure`, so it isn't waiting for the Deferred.
  1. `test__pamConvErrors` also doesn't return the Deferreds which come back from its two `assertFailure` calls.
  1. Nor does `test_tryAuthEdgeCases`.
  1. Each of the three modified files should have the date in the copyright header updated to include 2009.

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_48"></a>|z3p removed owner|
|-|-|

I added a bunch of tests for SSHPrivateKeyDatabase and SSHProtocolChecker, and I believe I've addressed the other points.  Ready for review.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_49"></a>|@therve set owner to z3p|
|-|-|

1)
```
+        self._cancelLoginTimeout = self.clock.callLater(self.loginTimeout,
                                                     self.timeoutAuthentication)
```
Indentation is a bit weird on the second line.

2)
```
+        f= getattr(self,'auth_%s' % (kind,), None)
```
Please a space before the equal sign.

3) test_requestAvatarId, test_requestAvatarId_without_signature and test_requestAvatarId_with_not_enough_authentication are missing docstrings.

I think I'm mostly happy with the branch, and we should move on. Please fix the above points, make a buildbot run, and merge is the supported ones are green. Thanks!

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_50"></a>|z3p commented|
|-|-|

(In [26456]) remove space before equals sign

Refs [#2682](https://github.com/twisted/twisted/issues/2682)

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_51"></a>|z3p commented|
|-|-|

(In [26457]) fix indentation

Refs [#2682](https://github.com/twisted/twisted/issues/2682)

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_52"></a>|z3p commented|
|-|-|

(In [26458]) docstrings

Refs [#2682](https://github.com/twisted/twisted/issues/2682)

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_53"></a>|z3p set status to `closed`|
|-|-|

(In [26534]) Merge 'userauth-2682-7'

Author: z3p
Reviewers: exarkun, radix, therve

Fixes [#2682](https://github.com/twisted/twisted/issues/2682)

This updates the SSHv2 user authentication framework in Conch tohave a full
set of unittests.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_54"></a>|@exarkun commented|
|-|-|

[#3871](https://github.com/twisted/twisted/issues/3871) is a regression introduced by this change.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
